src/refs.c

```diff
@@ -221,10 +221,17 @@ int git_reference_lookup_resolved(
 	assert(ref_out && repo && name);
 
 	if ((error = reference_normalize_for_repo(normalized, repo, name, true)) < 0 ||
-	    (error = git_repository_refdb__weakptr(&refdb, repo)) < 0 ||
-	    (error = git_refdb_resolve(ref_out, refdb, normalized, max_nesting)) < 0)
+	    (error = git_repository_refdb__weakptr(&refdb, repo)) < 0)
 		return error;
 
+	if ((error = git_refdb_resolve(ref_out, refdb, normalized, max_nesting)) < 0) {
```

Comment:
Do we ever use this pattern elsewhere where we both return an error _and_ return meaningful data (especially data that needs to be freed?).  This feels super ambiguous to me.

I think that we should return non-negative in this case - even `0` vs `1` would be an improvement, IMHO.

A `bool *exists` in the signature could be helpful, but it doesn't really feel very ergonomic.

I don't see the removal of the ad-hoc function that you mentioned, maybe we should bundle this commit with its removal so that it's clearer how/why this is changing?

Reply:
Sorry, I was reviewing out of order.  Now I understand better.

My complaint holds - this pattern feels really unwieldy to me ...  are there ways that you think we might be able to improve it?

Reply:
> My complaint holds - this pattern feels really unwieldy to me ... are there ways that you think we might be able to improve it?

@ethomson Yeah, I totally get your point and expected such a comment to come up. Mostly because I hate the interface myself and find it really hard to use correctly, too. How about we just document it may also return the last symbolic reference pointing to a nonexistent branch and pushing the burden onto the caller to check whether the returned reference is a symbolic reference or a direct reference?

Reply:
Done. Honestly, the resulting code does look a lot cleaner on call-sites now. Thanks for raising this point!